src/Connection/StreamConnection.php

```diff
@@ -308,7 +308,18 @@ protected function write($buffer)
     public function read()
     {
         $socket = $this->getResource();
-        $chunk = fgets($socket);
+        while (true) {
+            try {
+                $this->select([$socket], [], []);
```

Comment:
Is there any reason why this isn't implemented using non-blocking streams instead? http://php.net/manual/en/function.stream-set-blocking.php

Reply:
predis cannot  using non-blocking, because that socket io read must in loop.if you want real non-blocking, please google php redis react-php.

non-blocking need io event loop.

Reply:
Imho you can use non-blocking stream even without event loop. You just need to loop synchronously, just like you did here with `while( true)`, and check `fgets(...) !== false`. The point is that in such synchronous loop, signals are dispatched properly, unlike blocking stream.

Reply:
Indeed an event loop is most likely not really needed in such cases (see the socket-based phpiredis connection backend in which [we used `socket_select()`](https://github.com/nrk/predis/blob/v1.1.1/src/Connection/PhpiredisSocketConnection.php#L310) to provide the same timeout behaviour of `stream_socket_client()`, Predis remains mainly blocking so it doesn't need to coordinate with other I/O events or libraries that plugs into a full fledged event loop.